### PR TITLE
Add Rails 4.2 support

### PR DESCRIPTION
**Pull request by chaymaeBZ**
Defining a custom binary type on Rails 4 is different than Rails 5 and 6. the binary type interfaces are different. 

Documentation for ActiveRecord 4.2 Binary Type : https://www.rubydoc.info/gems/activerecord/4.2.4/ActiveRecord/Type/Binary

### CODE REVIEW

File: lib/ulid/rails/type.rb [hunk 1]
```diff
@@ -1,12 +1,19 @@
...
+end
+
 module ULID
   module Rails
-    class Type < ActiveModel::Type::Binary
-      class Data < ActiveModel::Type::Binary::Data
-        alias_method :hex, :to_s
+    class Type < TYPE_BINARY
+      class Data < TYPE_BINARY::Data
+        alias hex to_s
```
Comment (bquorning) [inline lib/ulid/rails/type.rb:16]: I think the PR would be easier to read if changes to logic were in a separate commit from style changes. Also, I’d recommend adding a style guide / RuboCop config or similar before doing style changes 😄

File: lib/ulid/rails/type.rb [hunk 2]
```diff
@@ -15,28 +22,56 @@ def initialize(formatter = Formatter, validator = Validator)
...
+        def type_cast_for_database(value)
+          return if value.nil?
+
+          case adapter
+          when 'mysql2', 'sqlite3'
+            Data.new(@formatter.unformat(value))
+          when 'postgresql'
+            Data.new([@formatter.unformat(value)].pack('H*'))
+          end
+        end
```
Comment (bquorning) [inline lib/ulid/rails/type.rb:50]: Do I read correctly that these three methods are completely identical for Rails <5 and >=5, except that the method names are different (we can probably add `super` to `#assert_valid_value` to make them completely identical)?

Could/should we extract the duplicated logic somehow?

File: lib/ulid/rails/type.rb [hunk 3]
```diff
@@ -15,28 +22,56 @@ def initialize(formatter = Formatter, validator = Validator)
...
+        def assert_valid_value(value)
+          raise ArgumentError, "`#{value}` is not a ULID format" unless @validator.is_valid?(value)
+        end
+
+        def deserialize(value)
+          return nil if value.nil?
+
+          value = value.to_s if value.is_a?(Data)
+          value = value.unpack1('H*') if value.encoding == Encoding::ASCII_8BIT
+          value = value[2..-1] if value.start_with?('\\x')
```
Comment (bquorning) [inline lib/ulid/rails/type.rb:61]: Please see #8.

File: lib/ulid/rails.rb [hunk 4]
```diff
@@ -15,7 +20,10 @@ module Rails
       def ulid(column_name, primary_key: false, auto_generate: nil)
         attribute column_name, ULID::Rails::Type.new
 
-        auto_generate = primary_key ? primary_key : auto_generate
+        auto_generate = primary_key || auto_generate
+
+        self.primary_key = column_name.to_s if primary_key && ActiveRecord::VERSION::MAJOR < 5
```
[the remark this answers is not in the document]
Comment (chaymaeBZ) [reply lib/ulid/rails.rb:25]: in Rails 5+, you don't need to tell the model about the primary key. This information pulled from the schema.

File: lib/ulid/rails/type.rb (hunk 2 above)
Comment (chaymaeBZ) [reply lib/ulid/rails/type.rb:37]: I hope it looks good now 😬

File: lib/ulid/rails/type.rb [hunk 5]
```diff
@@ -15,21 +22,41 @@ def initialize(formatter = Formatter, validator = Validator)
...
+          validate!(value)
+        end
+
+        def deserialize(value)
+          cast_value_from_database(value)
+        end
+
+        def serialize(value)
+          cast_type_to_data(value)
+        end
```
Comment (bquorning) [inline lib/ulid/rails/type.rb:50]:
```suggestion
        alias_method :assert_valid_value, :_assert_valid_value
        alias_method :deserialize, :_deserialize
        alias_method :serialize, :_serialize
```

File: lib/ulid/rails/type.rb [hunk 6]
```diff
@@ -40,7 +67,15 @@ def serialize(value)
         end
       end
 
-      private
+      def cast_value_from_database(value)
```
Comment (bquorning) [inline lib/ulid/rails/type.rb:70]:
```suggestion
      def _deserialize(value)
```

File: lib/ulid/rails/type.rb [hunk 7]
```diff
@@ -15,21 +22,41 @@ def initialize(formatter = Formatter, validator = Validator)
...
+      end
+
+      private
+
+      def validate!(value)
+        raise ArgumentError, "`#{value}` is not a ULID format" unless @validator.is_valid?(value)
       end
 
-      def serialize(value)
+      def cast_type_to_data(value)
```
Comment (bquorning) [inline lib/ulid/rails/type.rb:59]:
```suggestion
      def _serialize(value)
```

File: lib/ulid/rails/type.rb [hunk 8]
```diff
@@ -15,21 +22,41 @@ def initialize(formatter = Formatter, validator = Validator)
...
+        end
+
+        def serialize(value)
+          cast_type_to_data(value)
+        end
+      end
+
+      private
+
+      def validate!(value)
```
Comment (bquorning) [inline lib/ulid/rails/type.rb:55]:
```suggestion
      def assert_valid_value!(value)
```

File: lib/ulid/rails/type.rb [hunk 9]
```diff
@@ -15,21 +22,41 @@ def initialize(formatter = Formatter, validator = Validator)
...
       end
 
-      def assert_valid_value(value)
-        raise ArgumentError, "`#{value}` is not a ULID format" unless @validator.is_valid?(value)
-      end
+      if ActiveRecord::VERSION::MAJOR < 5
+        def type_cast_from_user(value)
+          validate!(value)
+          super
+        end
```
Comment (bquorning) [inline lib/ulid/rails/type.rb:29]:
```suggestion
        def type_cast_from_user(value)
          _assert_valid_value(value)
          super
        end
```

File: lib/ulid/rails/type.rb [hunk 10]
```diff
@@ -15,21 +22,41 @@ def initialize(formatter = Formatter, validator = Validator)
...
 
-      def deserialize(value)
-        return nil if value.nil?
+        def type_cast_from_database(value)
+          cast_value_from_database(value)
+        end
```
Comment (bquorning) [inline lib/ulid/rails/type.rb:33]:
```suggestion
        alias_method :type_cast_from_database, :_deserialize
```

File: lib/ulid/rails/type.rb [hunk 11]
```diff
@@ -15,21 +22,41 @@ def initialize(formatter = Formatter, validator = Validator)
...
 
-        value = value.to_s if value.is_a?(Data)
-        value = value.unpack1("H*") if value.encoding == Encoding::ASCII_8BIT
-        value = value[2..-1] if value.start_with?("\\x")
+        def type_cast_for_database(value)
+          cast_type_to_data(value)
+        end
```
Comment (bquorning) [inline lib/ulid/rails/type.rb:37]:
```suggestion
        alias_method :type_cast_for_database, :_serialize
```

File: lib/ulid/rails/type.rb [hunk 12]
```diff
@@ -15,21 +25,29 @@ def initialize(formatter = Formatter, validator = Validator)
...
+        def type_cast_from_user(value)
+          _assert_valid_value(value)
+          super()
+        end
 
-        value.length == 32 ? @formatter.format(value) : super
+      else
+        alias_method :assert_valid_value, :_assert_valid_value
+        alias_method :deserialize, :_deserialize
+        alias_method :serialize, :_serialize
```
Comment (bquorning) [inline lib/ulid/rails/type.rb:41]: My bad, I am so sorry: We cannot call `alias_method` before the target method has been defined. So maybe let’s just stick to the explicit calls you had before, e.g.
```rb
def serialize(value)
  _serialize(value)
end
```

File: lib/ulid/rails/type.rb [hunk 13]
```diff
@@ -40,6 +50,22 @@ def serialize(value)
...
+      if ActiveRecord::VERSION::MAJOR < 5
+
+        alias_method :type_cast_from_database, :_deserialize
+        alias_method :type_cast_for_database, :_serialize
+
+        def type_cast_from_user(value)
+          _assert_valid_value(value)
+          super
+        end
+
```
Comment (bquorning) [inline lib/ulid/rails/type.rb:62]:
```suggestion
        alias_method :type_cast_from_database, :_deserialize
        alias_method :type_cast_for_database, :_serialize

        def type_cast_from_user(value)
          _assert_valid_value(value)
          super
        end
```

File: lib/ulid/rails/type.rb [hunk 14]
```diff
@@ -40,6 +50,22 @@ def serialize(value)
         end
       end
```
Comment (bquorning) [inline lib/ulid/rails/type.rb:51]: We can make these three methods private with
```
private :_deserialize
```
etc.

**Review by bquorning — APPROVED** [review]

**Pull request merged**